File: 100/128 - Longest Consecutive Sequence.java
import java.util.HashMap;

public class Solution {
    public int longestConsecutive(int[] nums) {
        HashMap<Integer, Integer> hm = new HashMap<Integer, Integer>();
        int ret = 0, temp, len;
        
        for (int num : nums) {
            if (hm.containsKey(num))
                continue;
            
            if (hm.containsKey(num - 1)) {
                temp = hm.get(num - 1);
                if (temp > 0)
                    continue;
                else {
                    len = 1 - temp;
                    hm.put(num, -len);
                    hm.put(num - len, len);
                    ret = Math.max(ret, len + 1);
                }
            }
            
            
            if (hm.containsKey(num + 1)) {
                temp = hm.get(num + 1);
                if (temp >= 0) {
                    if (hm.containsKey(num)) {
                        len = -hm.get(num) + temp + 1;
                        hm.put(num + hm.get(num), len);
                        hm.put(num + 1 + temp, -len);
                        ret = Math.max(ret, len + 1);
                    } else {
                        len = temp + 1;
                        hm.put(num, len);
                        hm.put(num + len, -len);
                        ret = Math.max(ret, len + 1);
                    }
                }
                continue;
            }
            
            if (! hm.containsKey(num)) {
                hm.put(num, 0);
                ret = Math.max(ret, 1);
            }
        }
        
        return ret;
    }
}

File: csharp/128.longest-consecutive-sequence.cs
/*
 * @lc app=leetcode id=128 lang=csharp
 *
 * [128] Longest Consecutive Sequence
 */

using System.Collections.Generic;

public class Solution {
    public int LongestConsecutive(int[] nums) {
        IDictionary<int, int> map = new Dictionary<int, int>(nums.Length);
        int ret = 0;
        foreach (int curr in nums) {
            if (map.ContainsKey(curr)) {
                continue;
            }
            if (map.ContainsKey(curr - 1)) {
                int val = map[curr - 1];
                if (val > 0) {
                    continue;
                }
                int len = -val + 1;
                map[curr] = -len;
                map[curr - len] = len;
                ret = Math.Max(ret, len + 1);
            }
            if (map.ContainsKey(curr + 1)) {
                int val = map[curr + 1], len;
                if (val < 0) {
                    continue;
                }
                if (map.ContainsKey(curr)) {
                    len = -map[curr] + val + 1;
                    map[curr + map[curr]] = len;
                    map[curr + 1 + val] = -len;
                } else {
                    len = val + 1;
                    map[curr] = len;
                    map[curr + len] = -len;
                }
                ret = Math.Max(ret, len + 1);
                continue;
            }
            if (!map.ContainsKey(curr)) {
                map[curr] = 0;
                ret = Math.Max(ret, 1);
            }
        }
        return ret;
    }
}

